backends/simplego/dotgeneral.go

```diff
@@ -223,38 +277,85 @@ func execDotGeneral(backend *Backend, node *Node, inputs []*Buffer, _ []bool) (*
 	output := backend.getBufferForShape(outputShape)
 	output.Zeros()
 
-	// Problem size (per example of the batch):
+	// Check if inputs were pre-blocked at graph build time (via BlockForDotGeneral node).
+	lhsNode := node.inputs[0]
+	rhsNode := node.inputs[1]
+	lhsBlocked := lhsNode.opType == backends.OpTypeBlockForDotGeneral
+	rhsBlocked := rhsNode.opType == backends.OpTypeBlockForDotGeneral
+
+	// Handle pre-blocked cases
+	switch {
+	case lhsBlocked && rhsBlocked:
+		// Both inputs are pre-blocked - most efficient path
+		lhsBlockData := lhsNode.data.(*blockForDotGeneralData)
+		rhsBlockData := rhsNode.data.(*blockForDotGeneralData)
+		if err := execDotGeneralWithBothBlocked(backend, lhs, rhs, lhsBlockData, rhsBlockData, params, output); err != nil {
+			backend.putBuffer(output)
+			return nil, err
+		}
+		return output, nil
+
+	case lhsBlocked:
+		// Only LHS is pre-blocked
+		lhsBlockData := lhsNode.data.(*blockForDotGeneralData)
+		if err := execDotGeneralWithGraphBlockedLHS(backend, lhs, rhs, lhsBlockData, params, output); err != nil {
+			backend.putBuffer(output)
+			return nil, err
+		}
+		return output, nil
+
+	case rhsBlocked:
+		// Only RHS is pre-blocked
+		blockData := rhsNode.data.(*blockForDotGeneralData)
+		if err := execDotGeneralWithGraphBlockedRHS(backend, lhs, rhs, blockData, params, output); err != nil {
+			backend.putBuffer(output)
+			return nil, err
+		}
+		return output, nil
+	}
+
+	// Neither input is pre-blocked, try other execution paths
+
+	// Try the direct path for small matrices in contract-last order.
+	// This skips transpose but has strided RHS access, so only beneficial for small matrices.
+	if execDotGeneralSmallMatMul(backend, lhs, rhs, params, output) {
+		return output, nil
+	}
+
+	// Select execution path based on problem size.
+	// For large matrices, the blocked (cache-tiled) algorithm is more efficient.
+	// For smaller matrices, the normalized path with simple loops is sufficient.
 	crossesSize := params.rhsCrossSize * params.lhsCrossSize
 	blockDim := 1 << DotGeneralTargetBlockLog2Dim[dtype]
 	blockSize := blockDim * blockDim
 	var err error
-	problemSize := smallProblemSize
+	execPath := normalizedPath
 	if crossesSize > 16*blockSize {
-		problemSize = largeProblemSize
+		execPath = blockedPath
 	}
-	if backend.dotGeneralForceProblemSize != unknownProblemSize {
-		problemSize = backend.dotGeneralForceProblemSize
+	if backend.dotGeneralForceExecutionPath != autoSelectPath {
```

Comment:
So, I created `dotGeneralForceProblemSize` to help test the different paths. Would you like to extend it to work with SmallMatMul as well ?

For LHS/RHS that are in "matmux order", you could also add it to the `checkPath` version, that checks all paths generate the same values.